### PR TITLE
Use integer division trick to get exact base64 length 

### PR DESCRIPTION
**Pull request by gfoidl**
...and avoid (explicit) modulo operation.

Produces [less machine code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgNIAO1wYGAybOB4YqADLyYAIVjYA1gAUIuARgERJDALwMA7EgDcXHk+79ho8cXJI5ChgBUITWVUAH0WbAAbMABXSI8AQUkAEwA1KIFkjwB5GIxePNUYSQBzDAALAAp5WXkCjCLSirQGYAgISN9cGCgGjW0YPVwAShcOMZ4GSPMShgg8yOKrBkrK6dLhusLisvLhhgB6BkYAKgYUe0mrw6OK8SVuqZ2KhjgGAHdxd+w/GpgSnoMZLCATGcwMco9GAAOgmPHmGEWFgA1NZVlsGs9ygwAKTHfYAQmsAAZ9gB+c4MEAMYmXa7cA5HUQMfBGWQoFnQcRgcrYKC4BgAM2gDDuDFg+Gw8mSPRocO4AkFKwRSKsJNG1HpV2INhpjmo8rkSuq0h6fUkWh0uhGhs4mq1PHWswt73UFoF1hVSyODxUKDdlsGBiMJjMkn1Dp4ipWlS9FjxvrU/SthmMpnM+0s6sNkztkeucDgLoDuAj+YAvjn4QslqiGMWNAKzqQ6fmbqLIfWYvhgICIErIhpWlaJLz+R8YABySKdbDJZItABEAB0oMvJIvDZWDfbroyGABJJWfOZCHqC6bvBg/XyyO4WE9gH6SCCyKLTJ9iBjFeYlcqG/d8BgVkoAAT1FCA5jyeoO3EERIhicRcEgiooAgd5DWjWMawsAA+W9oQAWWwBB0gQmANTbVD0PrGAr1yDBskFQjgOgUCAFEEDAGBeHTSRKmGfVDR1FYamGOMy24bdJjGNwRC/LwfBqAIghCFBwiiWJ4jEJI0gyLIxAY+pGl2cgTVqSRjKxFo2g6LozQDAYhko+k83zfcxV9J4mmxN4T2+X4FH+QFgSEUEwwhKEWmZVkZEpEDuTHAVhSgWDxWAqUUllXcHSdKDESWNE1hmYZKhiMSMRMl5kQYUh9iOU5ziEnKtSwuM1RpFy224ETaTlFr6Sw+Rul6RyrRtAbrjc7qphmWjXUbZZiIqaEABFhAAJWA7CCskGy1LGoNU1DcwWgRWbSnSyVpR6QSq2uLCJUymVUqzTr7quaaZp4QsG3dSTuuk772rrP6YCbWrW3c25O0kbte1S/spiHYARx5PkBU+adZ3nJdV3XTdJqubcPv3I8J1Pc9L2vCxlPvCmn0kF83xnCBP3EH8Yj/UmjiAkDwIwSCERgsV4MQhhkI7NCMKJyY2pwhh8JqIiSLIxCuu66irxdBgGKYlj+c47jeLDATmrbETzPEnCAcmbdyyAA===).

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Convert.cs [hunk 1]
```diff
@@ -2527,16 +2527,16 @@ private static unsafe int ConvertToBase64Array(char* outChars, byte* inData, int
 
         private static int ToBase64_CalculateAndValidateOutputLength(int inputLength, bool insertLineBreaks)
         {
-            long outlen = ((long)inputLength) / 3 * 4;          // the base length - we want integer division here.
-            outlen += ((inputLength % 3) != 0) ? 4 : 0;         // at most 4 more chars for the remainder
+            // the base length - we want integer division here, at most 4 more chars for the remainder
+            long outlen = ((long)inputLength + 2) / 3 * 4;
 
             if (outlen == 0)
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/Convert.cs:2533]: Worth adding an assert that outlen is either 0 or 4?

Comment (gfoidl) [reply src/libraries/System.Private.CoreLib/src/System/Convert.cs:2533]: `outlen` will be the actual base64 encoded length, so >= 4 iif inputLength > 0.
`Debug.Assert(outlen % 4 == 0)` would be possible, but I don't see any value in adding that assert.

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Convert.cs:2533]: Ah, I see

File: src/libraries/System.Private.CoreLib/src/System/Convert.cs [hunk 2]
```diff
@@ -2527,16 +2527,16 @@ private static unsafe int ConvertToBase64Array(char* outChars, byte* inData, int
...
                 return 0;
 
             if (insertLineBreaks)
             {
-                long newLines = outlen / base64LineBreakPosition;
-                if ((outlen % base64LineBreakPosition) == 0)
+                long newLines = Math.DivRem(outlen, base64LineBreakPosition, out long remainder);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Convert.cs:2538]:
```suggestion
                (long newLines, long remainder) = Math.DivRem(outlen, base64LineBreakPosition);
```
We are using the new overload everywhere else in CoreLib.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Convert.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Convert.cs]: Does returning a tuple generate good code on all platforms now? I think I remember that this wasn't always the case.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Convert.cs]: Yes, we have analyzed the codegen and converted number of DivRem use the tuple returning overload in https://github.com/dotnet/runtime/pull/45074

Comment (davidfowl) [reply src/libraries/System.Private.CoreLib/src/System/Convert.cs]: Oh this is good to know

**Review by jkotas — APPROVED** [review]

**Comment by jkotas** [comment]
Thank you!

**Pull request merged**